src/modules/flow/persistence/persistence.c

```diff
@@ -0,0 +1,558 @@
+persist_drange_packet_new(const struct persist_data *data)
+{
+    struct persist_drange_data *mdata =
+        (struct persist_drange_data *)data;
+
+    return sol_flow_packet_new_drange(&mdata->last_value);
+}
+
+static int
+persist_drange_open(struct sol_flow_node *node,
+    void *data,
+    const struct sol_flow_node_options *options)
+{
+    struct persist_drange_data *mdata = data;
+    const struct sol_flow_node_type_persistence_float_options *opts =
+        (const struct sol_flow_node_type_persistence_float_options *)options;
+
+    mdata->base.packet_data_size = sizeof(struct sol_drange);
+    mdata->base.value_ptr = &mdata->last_value;
+    mdata->base.packet_new_fn = persist_drange_packet_new;
+    mdata->base.packet_data_get_fn = persist_drange_packet_data_get;
+    mdata->base.packet_send_fn = persist_drange_packet_send;
+    mdata->base.node_get_default_fn = persist_drange_node_get_default;
+    mdata->default_value = opts->default_value;
+
+    return persist_open(node, data, opts->storage, opts->name);
+}
+
+struct persist_string_data {
+    struct persist_data base;
+    const char *last_value;
+    char *default_value;
+};
+
+static void *
+persist_string_node_get_default(struct sol_flow_node *node)
+{
+    struct persist_string_data *mdata = sol_flow_node_get_private_data(node);
+
+    return mdata->default_value;
+}
+
+static int
+persist_string_packet_data_get(const struct sol_flow_packet *packet,
+    void *value_ptr)
+{
+    int r = sol_flow_packet_get_string(packet, value_ptr);
+
+    SOL_INT_CHECK(r, < 0, r);
+
+    return r;
+}
+
+static int
+persist_string_packet_send(struct sol_flow_node *node)
+{
+    struct persist_data *mdata = sol_flow_node_get_private_data(node);
+
+    return sol_flow_send_string_packet
+               (node, SOL_FLOW_NODE_TYPE_PERSISTENCE_STRING__OUT__OUT,
+               (const char *)mdata->value_ptr);
+}
+
+static struct sol_flow_packet *
+persist_string_packet_new(const struct persist_data *data)
+{
+    return sol_flow_packet_new_string(data->value_ptr);
+}
+
+static int
+persist_string_open(struct sol_flow_node *node,
+    void *data,
+    const struct sol_flow_node_options *options)
+{
+    struct persist_string_data *mdata = data;
+    const struct sol_flow_node_type_persistence_string_options *opts =
+        (const struct sol_flow_node_type_persistence_string_options *)options;
+
+    mdata->base.packet_new_fn = persist_string_packet_new;
+    mdata->base.packet_data_get_fn = persist_string_packet_data_get;
+    mdata->base.packet_send_fn = persist_string_packet_send;
+    mdata->base.node_get_default_fn = persist_string_node_get_default;
+
+    if (opts->default_value)
+        mdata->default_value = strdup((char *)opts->default_value);
+
+    return persist_open(node, data, opts->storage, opts->name);
+}
+
+static void
+persist_string_close(struct sol_flow_node *node, void *data)
+{
+    struct persist_string_data *mdata = data;
+
+    free(mdata->default_value);
```

Comment:
Is default_value properly initialized when not passed down from the opts struct?